### PR TITLE
fix: don't include replied mentions in mention filter

### LINKED ISSUE #2015: Subtract replies in mentions rule
The purpose of the mentions rule is to catch people spamming user mentions, but `message.mentions` shows the user being replied to, if the reply is pinging, which is out of scope for what the rule is trying to catch.

The replies should be subtracted when counting the number of mentions, and messages with only a reply but no mentions should not be returned.

### PR DESCRIPTION
**Pull request by onerandomusername**
closes #2015

- no longer counts mention replies
- no longer counts self-mentions

### CODE REVIEW

File: bot/rules/mentions.py [hunk 1]
```diff
@@ -15,10 +17,29 @@ async def apply(
...
+            continue
+        resolved = msg.reference.resolved
+        if isinstance(resolved, DeletedReferencedMessage):
+            # can't figure out the author
+            continue
+        if not resolved:
+            ref = msg.reference
+            resolved = await bot.instance.get_partial_messageable(ref.channel_id).fetch_message(ref.message_id)
+        if resolved.author in msg.mentions:
+            total_recent_mentions -= 1
```
Comment (mbaruh) [inline bot/rules/mentions.py:41]: It could be that the user is actually mentioned in the message though.
I may have set you on the wrong direction by saying "subtract", but it might be simpler to find all user mentions with regex.

Comment (onerandomusername) [reply bot/rules/mentions.py:41]: As discussed on [discord](https://discord.com/channels/267624335836053506/635950537262759947/921471985643569192), we're keeping this subtraction

File: bot/rules/mentions.py [hunk 2]
```diff
@@ -1,24 +1,65 @@
...
+        if not resolved:
+            ref = msg.reference
+            try:
+                # its possible, in a very unusual situation, for a message to have a reference
+                # that is both not in the cache, and deleted while running this function.
+                resolved = await bot.instance.get_partial_messageable(ref.channel_id).fetch_message(ref.message_id)
+            except NotFound:
+                log.info('Could not fetch the replied reference as its been deleted.')
+                continue
+        if resolved.author in msg.mentions:
```
Comment (TizzySaurus) [inline bot/rules/mentions.py:60]: This entire section could do with more blanklines separating the different if statements etc. for readability
```py
if not total_recent_mentions:
    ...
    
if message.type != MessageType.reply and not message.reference:
    ...
    
if isinstance(...):
    ...
    
# etc.
```

File: bot/rules/mentions.py [hunk 3]
```diff
@@ -1,24 +1,65 @@
...
         for user in msg.mentions
     )
 
+    # we don't want to include the mentions that the author mentioned with a reply
+    for msg in relevant_messages:
+        # break the loop to save processing if total_recent_mentions is 0
+        # additionally, ensures that we don't find a mention that wasn't originally there.
+        if not total_recent_mentions:
+            break
+        if msg.type != MessageType.reply and not msg.reference:
```
Comment (TizzySaurus) [inline bot/rules/mentions.py:45]: Should the first part of this if statement be an `isinstance`?

**Review by TizzySaurus — CHANGES_REQUESTED** [review]
Initial review

File: bot/rules/mentions.py (hunk 3 above)
Comment (onerandomusername) [reply bot/rules/mentions.py:45]: MessageType is an enum, is that still possible to ensure its the certain value of the enum?

File: bot/rules/mentions.py [hunk 4]
```diff
@@ -1,24 +1,65 @@
 from typing import Dict, Iterable, List, Optional, Tuple
 
-from discord import Member, Message
+from discord import DeletedReferencedMessage, Member, Message, MessageType, NotFound
+
+import bot
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply bot/rules/mentions.py:5]: > I'm not a fan of importing the entire project.


Same. Although it is how the instance is passed around everywhere so...

Its used here.
https://github.com/python-discord/bot/blob/5ca6be08a3ea47f124c204c3d972ee8d9704b707/bot/rules/mentions.py#L56

File: bot/rules/mentions.py [hunk 5]
```diff
@@ -1,24 +1,65 @@
...
-
+    # We use msg.mentions here as that is supplied by the api itself, to determine who was mentioned
+    # Additionally, msg.mentions includes the user replied to, even if the mention doesn't occur in the body.
+    # In order to exclude users who are mentioned as a reply, we check if the msg has a reference
+    #
+    # While we could use regex to parse the message content, and get a list of the mentions,
+    # that is very prone to breaking, as both the markdown, if discord used it, code block, etc
     total_recent_mentions = sum(
         not user.bot
+        and msg.author != user
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply bot/rules/mentions.py:34]: I think this is more readable than if its moved.

File: bot/rules/mentions.py [hunk 6]
```diff
@@ -1,24 +1,65 @@
...
         for msg in recent_messages
         if msg.author == last_message.author
     )
-
+    # We use msg.mentions here as that is supplied by the api itself, to determine who was mentioned
+    # Additionally, msg.mentions includes the user replied to, even if the mention doesn't occur in the body.
+    # In order to exclude users who are mentioned as a reply, we check if the msg has a reference
+    #
+    # While we could use regex to parse the message content, and get a list of the mentions,
+    # that is very prone to breaking, as both the markdown, if discord used it, code block, etc
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply bot/rules/mentions.py]: https://github.com/python-discord/bot/pull/2017/commits/c2bbb9c10ed1e9172a51cbb2780676ddb9fd77c1

File: bot/rules/mentions.py [hunk 7]
```diff
@@ -1,24 +1,65 @@
...
     total_recent_mentions = sum(
         not user.bot
+        and msg.author != user
         for msg in relevant_messages
         for user in msg.mentions
     )
 
+    # we don't want to include the mentions that the author mentioned with a reply
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply bot/rules/mentions.py]: https://github.com/python-discord/bot/pull/2017/commits/c2bbb9c10ed1e9172a51cbb2780676ddb9fd77c1

File: bot/rules/mentions.py [hunk 8]
```diff
@@ -1,24 +1,65 @@
...
+    for msg in relevant_messages:
+        # break the loop to save processing if total_recent_mentions is 0
+        # additionally, ensures that we don't find a mention that wasn't originally there.
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply bot/rules/mentions.py]: https://github.com/python-discord/bot/pull/2017/commits/c2bbb9c10ed1e9172a51cbb2780676ddb9fd77c1

File: bot/rules/mentions.py [hunk 9]
```diff
@@ -1,24 +1,65 @@
...
+        if msg.type != MessageType.reply and not msg.reference:
+            continue
+        resolved = msg.reference.resolved
+        if isinstance(resolved, DeletedReferencedMessage):
+            # can't figure out the author
+            continue
+        if not resolved:
+            ref = msg.reference
+            try:
+                # its possible, in a very unusual situation, for a message to have a reference
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply bot/rules/mentions.py]: https://github.com/python-discord/bot/pull/2017/commits/c2bbb9c10ed1e9172a51cbb2780676ddb9fd77c1

File: bot/rules/mentions.py (hunk 2 above)
Comment (onerandomusername) [reply bot/rules/mentions.py]: https://github.com/python-discord/bot/pull/2017/commits/c2bbb9c10ed1e9172a51cbb2780676ddb9fd77c1 okay yea bad commit message

File: bot/rules/mentions.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TizzySaurus) [reply bot/rules/mentions.py:5]: Hmm, I see.

In which case that can stay, but I would still like to see a `from bot.log import get_logger`, which is also done in [other instances of this scenario](https://sourcegraph.com/github.com/python-discord/bot/-/blob/bot/exts/help_channels/_message.py?L11)

File: bot/rules/mentions.py [hunk 10]
```diff
@@ -1,24 +1,72 @@
...
         if msg.author == last_message.author
     )
-
+    # We use msg.mentions here as that is supplied by the api itself, to determine who was mentioned
+    # Additionally, msg.mentions includes the user replied to, even if the mention doesn't occur in the body.
+    # In order to exclude users who are mentioned as a reply, we check if the msg has a reference
+    #
+    # While we could use regex to parse the message content, and get a list of
+    # the mentions, that solution is very prone to breaking.
+    # We would need to deal with codeblocks, escaping markdown. and any discrepancies between
```
Comment (TizzySaurus) [inline bot/rules/mentions.py:33]:
```suggestion
    # We would need to deal with codeblocks, escaping markdown and any discrepancies between
```

File: bot/rules/mentions.py (hunk 3 above)
Comment (TizzySaurus) [reply bot/rules/mentions.py:49]: Right, this is correct -- an `isinstance` would likely give an error.

File: bot/rules/mentions.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TizzySaurus) [reply bot/rules/mentions.py:37]: I was thinking moreso because the thing you're summing is `not user.bot and msg.author != user`, and so should go all on one line. The "extra" lines in `sum` are for what you're iterating over; not for expanding the condition.

Also, if you do move to one line, then `not (user.bot or msg.author == user)` is better in terms of efficiency, but it's arguably not worth the readability downgrade so will leave up to you.

**Review by TizzySaurus — CHANGES_REQUESTED** [review]
There's a random `.` in a docstring, and more discussion about the `sum()`

File: bot/rules/mentions.py (hunk 10 above)
Comment (onerandomusername) [reply bot/rules/mentions.py:33]: should be a comma, will fix.

Comment (onerandomusername) [reply bot/rules/mentions.py]: https://github.com/python-discord/bot/pull/2017/commits/95d654cd27f93d8673affd295b2f197f37089da0

**Review by Bluenix2 — APPROVED** [review]
I want to apologize to the staff that I pinged while attempting to test this... that said thank you for the PR, works nicely!

**Review by wookie184 — DISMISSED** [review]
The logic here isn't quite correct, as this will currently subtract 1 from the total even if the reply was to a bot or themself, in which case it wasn't counted in the first place.

**Comment by ichard26** [comment]
Hi @onerandomusername, do you still want to work on this PR or should I put this up for grabs? I kinda want to work on this hence why I'm triaging this :p

**Comment by wookie184** [comment]
@ichard26 if you're still interested in this feel free to assign yourself, otherwise this can be marked as up for grabs.

**Comment by ichard26** [comment]
No, not really anymore. I'll put it up for grabs.

**Comment by TizzySaurus** [comment]
> The logic here isn't quite correct, as this will currently subtract 1 from the total even if the reply was to a bot or themself, in which case it wasn't counted in the first place.

This has been fixed (I took over the PR) :+1:

File: bot/rules/mentions.py [hunk 11]
```diff
@@ -1,24 +1,75 @@
...
+    In very rare cases, may not be able to determine a
+    mention was to a reply, in which case it is not ignored.
+    """
     relevant_messages = tuple(
         msg
         for msg in recent_messages
         if msg.author == last_message.author
     )
-
+    # We use `msg.mentions` here as that is supplied by the api itself, to determine who was mentioned
```
[the remark this answers is not in the document]
Comment (TizzySaurus) [reply bot/rules/mentions.py]: Fixed :+1:

**Comment by TizzySaurus** [comment]
I've just fixed the tests for this PR, but am looking to also add a new test that ensures this PR actually works (pings from replies don't work).

This means creating a new `MockReferencedMessage` class which I don't have the time to look into today, but will hopefully be done over the upcoming days.

**Comment by TizzySaurus** [comment]
This PR is now ready for review again 😄

**Review by wookie184 — COMMENTED** [review]
Thanks for taking this over and fixing it up 👍 

I have one comment, other than that seems alright, although I do think there is room for refactoring that could make it a bit cleaner. Given this is quite an old PR now i'd be happy if you'd prefer to leave it as is, although these are my thoughts:

It would be clearer to just have a loop that does all the checks necessary before incrementing the total. The logic to get the replied to message is also quite involved, so could be moved to a separate function, possibly in `bot.utils.messages`. You'd then have something more like this:

```python
total_mentions = 0
for msg in relevant_messages:
    replied_to_msg = utils.messages.get_replied_to_message(msg)
    for user in msg.mentions:
        # Don't count bot messages or self mentions
        if user.bot or user == msg.author:
            continue
        # Don't count mentions of the user being replied to
        if replied_to_msg and replied_to_msg.author == user:
            continue
        total_mentions += 1
```

File: bot/rules/mentions.py [hunk 12]
```diff
@@ -1,24 +1,77 @@
...
     )
 
+    # No reason to run processing and fetch messages if there are no mentions.
+    if not total_recent_mentions:
+        return None
+
+    # We don't want to include mentions that are to the replied user in message replies.
+    for msg in relevant_messages:
+
+        if msg.type != MessageType.reply and not msg.reference:
```
[the remark this answers is not in the document]
Comment (TizzySaurus) [reply bot/rules/mentions.py:48]: I wasn't sure myself so asked in the dpy server.

Turns out the difference is that `msg.reference` exists for both replies *and* the system message you get from pinning a message.

`MessageType.reply` will only be for replies, and *not* the system message from pins.

TL;DR; This can be `if msg.type != MessageType.reply` since we don't care about the pin system message.

**Comment by TizzySaurus** [comment]
> It would be clearer to just have a loop that does all the checks necessary before incrementing the total. The logic to get the replied to message is also quite involved, so could be moved to a separate function, possibly in `bot.utils.messages`. You'd then have something more like this:
> ...

Yeah, I completely agree. The current logic was done by @onerandomusername and I didn't think to adjust it. I'll do this :+1:
